### PR TITLE
[feature-wip](array-type) add function arrays_overlap

### LINKED ISSUE #10052: [Feature] Add lots of Array Functions
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Description

We have supported `array` data type recently, all pr(s) about `array` please refer to https://github.com/apache/incubator-doris/issues/7570 .

Now we need to add lots of array functions to make `array` more powerful.
If anybody is interesting to `array`, please let me know.

Array Functions DONE:
- size/cardinality
- array_position
- array_contains
- element_at

Array Functions DOING:
- array_sum
- array_avg
- array_min
- array_max
- array_product
- array_distinct
- array_remove
- arrays_overlap

Array Functions TODO:
- array_union
- array_intersect
- array_except
- reverse
- concat_ws
- array_join
- array_sort
- array_slice

### Use case

_No response_

### Related issues

https://github.com/apache/incubator-doris/issues/7570

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by cambyzju**
# Proposed changes

Issue Number: close #10052

## Problem Summary:
arrays_overlap check if there is any common element for left and right array.

Usage example:
```
mysql> set enable_vectorized_engine=true;
mysql> select c_left,c_right,arrays_overlap(c_left,c_right) from array_test;
+--------------+-----------+-------------------------------------+
| c_left       | c_right   | arrays_overlap(`c_left`, `c_right`) |
+--------------+-----------+-------------------------------------+
| [1, 2, 3]    | [3, 4, 5] |                                   1 |
| [1, 2, 3]    | [5, 6]    |                                   0 |
| [1, 2, NULL] | [1]       |                                NULL |
| NULL         | [1, 2]    |                                NULL |
| [1, 2, 3]    | [1, 2]    |                                   1 |
+--------------+-----------+-------------------------------------+
```

## Checklist(Required)

1. Does it affect the original behavior: (Yes/No/I Don't know)
2. Has unit tests been added: (Yes/No/No Need)
3. Has document been added or modified: (Yes/No/No Need)
4. Does it need to update dependencies: (Yes/No)
5. Are there any changes that cannot be rolled back: (Yes/No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/vec/functions/array/function_arrays_overlap.h [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+            if (dst_nullmap_data[row]) {
+                continue;
+            }
+
+            ssize_t left_start = (*left_data.offsets_ptr)[row - 1];
+            ssize_t left_size = (*left_data.offsets_ptr)[row] - left_start;
+            ssize_t right_start = (*right_data.offsets_ptr)[row - 1];
+            ssize_t right_size = (*right_data.offsets_ptr)[row] - right_start;
+            if (left_size == 0 || right_size == 0) {
+                dst_data[row] = 0;
```
Comment (HappenLee) [inline be/src/vec/functions/array/function_arrays_overlap.h:231]: `continue`

File: be/src/vec/functions/array/function_arrays_overlap.h [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+            }
+
+            ExecutorImpl impl;
```
Comment (HappenLee) [inline be/src/vec/functions/array/function_arrays_overlap.h:234]: chose size lower case to build hash set to opt the case

File: be/src/vec/functions/array/function_arrays_overlap.h [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+            impl.insert_array(left_data.nested_col, left_start, left_size);
+
+            if (impl.find_any(right_data.nested_col, right_start, right_size)) {
```
Comment (HappenLee) [inline be/src/vec/functions/array/function_arrays_overlap.h:237]: `dst_data[row] = impl.find_any(right_data.nested_col, right_start, right_size)`

File: be/src/vec/functions/array/function_arrays_overlap.h (hunk 1 above)
Comment (cambyzju) [reply be/src/vec/functions/array/function_arrays_overlap.h:231]: done

File: be/src/vec/functions/array/function_arrays_overlap.h (hunk 2 above)
Comment (cambyzju) [reply be/src/vec/functions/array/function_arrays_overlap.h:235]: done

File: be/src/vec/functions/array/function_arrays_overlap.h (hunk 3 above)
Comment (cambyzju) [reply be/src/vec/functions/array/function_arrays_overlap.h]: done

File: be/src/vec/functions/array/function_arrays_overlap.h [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+            if (data.array_nullmap_data && data.array_nullmap_data[row]) {
+                dst_nullmap_data[row] = 1;
+                continue;
+            }
+
+            // any element inside array is NULL, return NULL
+            ssize_t start = (*data.offsets_ptr)[row - 1];
+            ssize_t size = (*data.offsets_ptr)[row] - start;
+            for (ssize_t i = start; i < start + size; ++i) {
+                if (data.nested_nullmap_data && data.nested_nullmap_data[i]) {
```
[the remark this answers is not in the document]
Comment (cambyzju) [reply be/src/vec/functions/array/function_arrays_overlap.h]: done

**Review by HappenLee — DISMISSED** [review]
LGTM

**Comment by cambyzju** [comment]
rebase to trigger regression test again

**Comment by cambyzju** [comment]
rebase to trigger FE UT and regression test again

**Review by yangzhg — APPROVED** [review]
LGTM

**Comment by 520lailai** [comment]
why arrays_overlap([1, 2, NULL]  , [1] ) will return NULL ?  This is not logical, in starRock it will return true.

**Pull request merged**